core/storage/btree.rs

```diff
@@ -484,136 +604,207 @@ impl BTreeCursor {
+                    // we can consider dividers the first cell of each page starting from the second page
+                    for (page_id_index, page) in
+                        new_pages.iter_mut().take(new_pages_len - 1).enumerate()
+                    {
+                        assert!(page.cell_count() > 1);
+                        let divider_cell_index = divider_cells_index[page_id_index];
+                        let cell_payload = scratch_cells[divider_cell_index];
+                        let cell = read_btree_cell(cell_payload, &page.page_type(), 0).unwrap();
+                        if is_leaf {
+                            // create a new divider cell and push
+                            let key = match cell {
+                                BTreeCell::TableLeafCell(leaf) => leaf._rowid,
+                                _ => unreachable!(),
+                            };
+                            let mut divider_cell = Vec::new();
+                            divider_cell.extend_from_slice(
+                                &(new_pages_ids[page_id_index] as u32).to_be_bytes(),
+                            );
+                            divider_cell.extend(std::iter::repeat(0).take(9));
+                            let n = write_varint(&mut divider_cell.as_mut_slice()[4..], key);
+                            divider_cell.truncate(4 + n);
+                            let parent_cell_idx = find_cell(parent, key);
+                            self.insert_into_cell(parent, divider_cell.as_slice(), parent_cell_idx);
+                        } else {
+                            // move cell
+                            let key = match cell {
+                                BTreeCell::TableInteriorCell(interior) => interior._rowid,
+                                _ => unreachable!(),
+                            };
+                            let parent_cell_idx = find_cell(page, key);
+                            self.insert_into_cell(parent, cell_payload, parent_cell_idx);
+                            // self.drop_cell(*page, 0);
+                        }
+                    }
+
+                    {
+                        // copy last page id to right pointer
+                        let last_pointer = *new_pages_ids.last().unwrap() as u32;
+                        parent.write_u32(right_pointer, last_pointer);
+                    }
                 }
+            }
+
+            self.page = RefCell::new(Some(mem_page.parent.as_ref().unwrap().clone()));
+        }
+    }
 
-                break;
+    fn balance_root(&mut self) {
+        /* todo: balance deeper, create child and copy contents of root there. Then split root */
+        /* if we are in root page then we just need to create a new root and push key there */
+        let mem_page = {
+            let mem_page = self.page.borrow();
+            let mem_page = mem_page.as_ref().unwrap();
+            mem_page.clone()
+        };
```

Comment:
Maybe we want a helper function for this sequence of code because it keeps appearing in multiple places

Reply:
done